blackbird_python/blackbird/tests/test_listener.py

```diff
@@ -0,0 +1,464 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for the Blackbird parser/listener"""
+# pylint: disable=too-many-ancestors,no-self-use,redefined-outer-name,too-many-arguments,no-value-for-parameter,too-few-public-methods
+import pytest
+
+import numpy as np
+
+import antlr4
+
+from blackbird.blackbirdLexer import blackbirdLexer
+from blackbird.blackbirdParser import blackbirdParser
+from blackbird.listener import BlackbirdListener, parse_blackbird
+
+
+test_file = """
+# a test blackbird script
+
+float alpha = 0.3423
+
+with fock(num_subsystems=1, cutoff_dim=7, shots=10):
+    Coherent(alpha, sqrt(pi)) | 0
+    MeasureFock() | 0
+"""
+
+
+@pytest.fixture
+def device():
+    """Mock out the device inside the blackbird parser"""
+    class DummyDeviceContext:
+        """Dummy device context with mocked out device name"""
+        getText = lambda self: "MockedDevice"
+    return DummyDeviceContext
+
+
+@pytest.fixture
+def parser():
+    """Create a parser for the test"""
+    def _parser(text):
+        """The parser fixture accepts a blackbird string to parse"""
+        lexer = blackbirdLexer(antlr4.InputStream(text))
+        stream = antlr4.CommonTokenStream(lexer)
+        parser = blackbirdParser(stream)
+        return parser
+    return _parser
+
+
+class TestParsingVariables:
+    """Tests for parsing variable declarations"""
+
+    def test_integer_variable(self, parser, device, monkeypatch):
+        """Test that an integer variable is correctly parsed"""
+        parser = parser("int n = 5")
+        tree = parser.start()
```

Comment:
A lot of this code is duplicated. Can we put this stuff in a separate function in this file?